### PR TITLE
Add msys2/mingw CI build for the external libraries

### LINKED ISSUE #29: OpenModelica
Hi all,

I read somethings about the OpenModelica troubles. My question, is the library working in openmodelica at all? Or is it just the examples that don't work?

kind regards

### PR DESCRIPTION
**Pull request by adrpo**
This PR adds a CI build for the external libraries on mingw. [It previously just added the  binaries directly].

Partially fixes #29.

### CODE REVIEW

**Comment by tbeu** [comment]
Thanks for the PR. As libexpat was updated by (8a68ebd9abede31f60a247fb01584082ec7fc390) I'd like to ask you if you could add a Appveyor CI to build the mingw32/64 libs there. Having it as integrated CI would be of greater value than just updating the libs occasionally. Thanks again.

**Comment by adrpo** [comment]
Sure. I don't have too much experience with Appveyor CI but I'll see what I can do.

**Comment by tbeu** [comment]
I think I can provide teh initial setup. Would be good to have it before end of month such that i can update the MoCo paper. Thanks.

**Comment by adrpo** [comment]
I will try using msys2 package as explained here: https://docs.travis-ci.com/user/reference/windows/#how-do-i-use-msys2.
Will update my PR with it soon.

**Comment by tbeu** [comment]
Initial setup might be tricky: Have a look at https://github.com/tbeu/matio/blob/master/.appveyor.yml#L133

**Comment by tbeu** [comment]
Please rebase your branch on current master and note that libED_SSVFile was replaced by libED_XML2File with new dependency on libxml2. Thanks.

**Comment by adrpo** [comment]
This needs more work but I will keep working on in.

**Comment by adrpo** [comment]
At least it builds now, I'll clean it up tomorrow or so and then request a review.

File: .travis.yml [hunk 1]
```diff
@@ -20,46 +20,112 @@ env:
...
+        popd
+        cd ExternData/Resources/Library/$PLATFORM
+        tar cJf ExternData_$PLATFORM.tar.xz $DEPLOY_LIBS
+        if [[ "$CC" == "gcc-4.8" ]]; then sh ../../../../upload-to-bitbucket.sh tbeu $BBPASS /tbeu/downloads/downloads ExternData_$PLATFORM.tar.xz; fi
+        ;;
+      osx)
+        ## nothing yet
+        ;;
+      windows)
+        $mingw64 ./.msys2-build.sh && ./.msys2-build.sh mingw64
```
Comment (tbeu) [inline .travis.yml]: * Can we please have two configurations in the build matrix?
* Should targetdir always be set?
* Should targetdir be win32 and win64 where binaries go to compiler specific subdir?

**Comment by tbeu** [comment]
@adrpo I am going to release v3.0.0 latest by Friday (30/7) night. Would be good to have your build script CI merged. Thanks.

**Comment by adrpo** [comment]
Agreed, will play with it tomorrow.

File: .travis.yml (hunk 1 above)
Comment (adrpo) [reply .travis.yml]: Done.

Comment (adrpo) [reply .travis.yml]: For mingw we should have mingw64/mingw32 for Visual Studio we should have win32 and win64.

**Comment by tbeu** [comment]
Some issues with paths (tar and upload).

**Comment by tbeu** [comment]
Can we put the binaries to Library/win32/gcc810 and Library/win64/gcc810 which would be compliant with MLS 3.4?

**Comment by adrpo** [comment]
Seems to be OK with the spec: https://specification.modelica.org/maint/3.5/functions.html#annotations-for-external-libraries-and-include-files
I am unsure if we support this layout in OpenModelica but I will check. I know that we support mingw64 and mingw32 at the same level with win32 and win64.

**Comment by tbeu** [comment]
> I am unsure if we support this layout in OpenModelica but I will check. I know that we support mingw64 and mingw32 at the same level with win32 and win64.

I can try with symlinks.

**Comment by tbeu** [comment]
Seems, the Travis CI requires more credits. I already asked their support (as I now usually need to do).

**Comment by adrpo** [comment]
Hmm, I guess I hit some sort of quota for builds as the last build was cancelled.

**Comment by adrpo** [comment]
Ah, OK, i now saw your comment. I guess we can wait with it until tomorrow or so.

**Comment by tbeu** [comment]
Yep, just need to wait to see some more credits

image

**Comment by adrpo** [comment]
Seems the builds are fine now but their upload to bitbucket fails.

```
    [!] error: missing arguments :(
     syntax: upload-to-bitbucket.sh <user> <password> <repo downloads page> <local file to upload>
    example: upload-to-bitbucket.sh swyter secret1 /Swyter/tld-downloads/downloads myfile.zip
    tip: if a file with the same name is already in there it will be overwrited without notice
```

I have no idea how to fix this one.

**Comment by adrpo** [comment]
For mingw i actually get:
```
actual upload progress should appear right now as a progress bar, be patient:
[!] server error: bitbucket (the platform) returned HTTP error code 000
```
from the upload script.

**Comment by tbeu** [comment]
There is another error:

```
Moving libxml2.a
+mv libxml2.a ../../ExternData/Resources/Library/linux64/libxml2.a
mv: cannot stat 'libxml2.a': No such file or directory
+popd
./.build.sh: 27: ./.build.sh: popd: not found
+echo Moving libxml2.a
Moving libxml2.a
```

**Comment by tbeu** [comment]
> Seems the builds are fine now but their upload to bitbucket fails.
> 
> ```
>     [!] error: missing arguments :(
>      syntax: upload-to-bitbucket.sh <user> <password> <repo downloads page> <local file to upload>
>     example: upload-to-bitbucket.sh swyter secret1 /Swyter/tld-downloads/downloads myfile.zip
>     tip: if a file with the same name is already in there it will be overwrited without notice
> ```
> 
> I have no idea how to fix this one.

I can only assume that pathes do not match.

**Comment by tbeu** [comment]
Can you please force push your changes again. I did not notice that you commited when I triggered the build again. Sorry for that,

**Comment by adrpo** [comment]
I think I found what the problem is, the build is building in a specific directory and the upload script is called from the top level, will fix it.

**Comment by adrpo** [comment]
You already fixed it, I have some other way of fixing it without the cd, basically using the path to the upload file:
`$mingw32 ./upload-to-bitbucket.sh tbeu $BBPASS /tbeu/downloads/downloads ./ExternData/Resources/Library/$PLATFORM/ExternData_$PLATFORM.tar.xz`

Should I push my change or just leave yours?

**Comment by tbeu** [comment]
Hang on, I am about to fix it.

**Comment by adrpo** [comment]
Ok. Let me know if you need anything more.

**Comment by adrpo** [comment]
Note that you cannot do:

```
$mingw64 cd someplace
$mingw64 copy from someplace
```
as each command starts from the top again, it has no previous context, so do:
```
$mingw64 cd someplace && copy from someplace
```

**Comment by tbeu** [comment]
Can we use cmd or ps1 from https://bitbucket.org/swyter/bitbucket-curl-upload-to-repo-downloads/src/default/ for the uploads as it gives syntax errors otherwise?

**Comment by tbeu** [comment]
Here's my current setup: https://github.com/modelica-3rdparty/ExternData/tree/AddMinGWOpenModelicaLibs

**Comment by tbeu** [comment]
Sorry, I cancelled your build as we will run out of credits to soon then.

**Comment by adrpo** [comment]
No problem. I will check on my computer if is works fine on my local mingw.

**Comment by adrpo** [comment]
With the latest changes it seems to be working fine on my computer. You can try to run it again tomorrow, sorry I had so many tentative changes, is hard to test properly otherwise.

**Comment by tbeu** [comment]
It still gives "error: missing arguments :(" when uploading. I have no more clue,

**Comment by tbeu** [comment]
All merged. Thanks a lot!

**Comment by tbeu** [comment]
An this is today's usage statistics:

image

**Pull request merged**